core/src/main/scala/org/apache/spark/h2o/converters/PrimitiveRDDConverter.scala

```diff
@@ -57,25 +65,15 @@ private[converters] object PrimitiveRDDConverter extends Logging with ConverterU
   def perPrimitiveRDDPartition[T]() // extra arguments for this transformation
                                  (keyName: String, vecTypes: Array[Byte], uploadPlan: Option[immutable.Map[Int, NodeDesc]]) // general arguments
                                  (context: TaskContext, it: Iterator[T]): (Int, Long) = { // arguments and return types needed for spark's runJob input
-    val con = ConverterUtils.getWriteConverterContext(uploadPlan, context.partitionId())
 
+    val (iterator, dataSize) = WriteConverterCtxUtils.bufferedIteratorWithSize(uploadPlan, it)
+    val con = WriteConverterCtxUtils.create(uploadPlan, context.partitionId(), dataSize)
     con.createChunks(keyName, vecTypes, context.partitionId())
-    // try to wait for reply to ensure we can continue with sending
-    it.foreach { r =>
-      r match {
-        case n: Number => con.put(0, n)
-        case n: Boolean => con.put(0, n)
-        case n: String => con.put(0, n)
-        case n: java.sql.Timestamp => con.put(0, n)
-        case _ => con.putNA(0)
-      }
-      con.increaseRowCounter()
-    }
+    iterator.foreach {con.putAnySupportedType(0, _)}
     //Compress & write data in partitions to H2O Chunks
     con.closeChunks()
-
     // Return Partition number and number of rows in this partition
-    (context.partitionId, con.numOfRows)
+    (context.partitionId, iterator.size)
```

Comment:
Iterator in Scala is not something that a) guarantees to provide size (it can be unknown), b) is usable after sizing (TraversableOnce, e.g.) c) it may not even have a size.

If you have an iterator, using size may be destructive. If you know the size ahead, and it's not heavy, it may be better to use a different type. Seq, for instance.